parse_mod_info: Filter empty strings to marshmallow doesn't get angry

duk3luk3 opened the pull request:
marshmallow does not like empty strings in `UrlField`s and therefore failed validation if a `mod_info.lua` contained `url = ""`.

This commit removes all `None` and `""` values from the data before passing it to the marshmallow schema.

Fixes #31

Also contains a lua fix:

Ignore "internal" globals when parsing lua: When we execute lua scripts to extract the global variables they set, we are not interested in the things that already hang around in the globals table when the Lua VM is initialized.

Since these things also contain binary that we don't want to run through string decoding (resulting in `UnicodeDecodeError`), we're better off getting rid of them.

Sheeo commented:
Don't you just want to update [the schema](https://github.com/FAForever/faftools/blob/develop/faf/schema/mod_info.py#L14) with `url = fields.Url(default="", allow_none=True)`?

duk3luk3 commented:
Ah, I didn't know that was possible! I didn't find that in the marshmallow docs when I was looking.

duk3luk3 commented:
This doesn't actually work, and apparently the workaround is to remove all "empty" values before passing the data to marshmallow - https://github.com/marshmallow-code/marshmallow/issues/167#issuecomment-266838585

For a schema description tool this is kind of shite.

duk3luk3 committed:
Ignore "internal" globals when parsing lua

When we execute lua scripts to extract the global variables they set, we
are not interested in the things that already hang around in the globals
table when the Lua VM is initialized.

Since these things also contain binary that we don't want to run through
string decoding (resulting in `UnicodeDecodeError`), we're better off
getting rid of them.

duk3luk3 committed:
Fix mod validation failing for empty string values

marshmallow does not like empty strings in `UrlField`s and therefore
failed validation if a `mod_info.lua` contained `url = ""`.

This commit removes all `None` and `""` values from the data before
passing it to the marshmallow schema.

Fixes #31

duk3luk3 commented:
@Sheeo can you reproduce that without my commit to ignore the "internal" lua globals using `from_lua` on anything causes a UnicodeDecodeError? I don't quite understand why the python-api doesn't already suffer from this problem.

duk3luk3 commented:
Here is what it takes for me to trigger the problem:

```python
from faf.tools.lua import parse
parse.from_lua('')
```

results in this:

```
(.venv) erlacher@laprbg12 faftools(develop|…) % PYTHONSTARTUP=~/.pythonrc python                                                                                                                                            ~/git/faf/faftools
Python 3.6.2 (default, Jul 20 2017, 03:52:27) 
[GCC 7.1.1 20170630] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from faf.tools.lua import parse
>>> parse.from_lua('')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/erlacher/git/faf/faftools/faf/tools/lua/parse.py", line 31, in from_lua
    return unfold_table(lua.globals())
  File "/home/erlacher/git/faf/faftools/faf/tools/lua/parse.py", line 28, in unfold_table
    result[k] = dict(v)
  File "lupa/_lupa.pyx", line 600, in lupa._lupa._LuaObject.__getitem__ (lupa/_lupa.c:10396)
  File "lupa/_lupa.pyx", line 617, in lupa._lupa._LuaObject._getitem (lupa/_lupa.c:10593)
  File "lupa/_lupa.pyx", line 1074, in lupa._lupa.py_from_lua (lupa/_lupa.c:16639)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc2 in position 4: invalid continuation byte
>>> 
```
